pallets/loans-ref/src/types.rs

```diff
@@ -0,0 +1,662 @@
+		ensure!(
+			amount <= self.max_borrow_amount()?,
+			Error::<T>::from(BorrowLoanError::MaxAmountExceeded)
+		);
+
+		ensure!(
+			self.info.schedule.maturity.date() > T::Time::now().as_secs(),
+			Error::<T>::from(BorrowLoanError::MaturityDatePassed)
+		);
+
+		match self.info.restrictions.borrows {
+			BorrowRestrictions::WrittenOff => {
+				ensure!(
+					self.written_off_status.is_none(),
+					Error::<T>::from(BorrowLoanError::WrittenOffRestriction)
+				)
+			}
+		}
+
+		Ok(())
+	}
+
+	fn ensure_can_repay(&self, amount: T::Balance) -> Result<T::Balance, DispatchError> {
+		let current_debt =
+			T::InterestAccrual::current_debt(self.info.interest_rate, self.normalized_debt)?;
+
+		match self.info.restrictions.repayments {
+			RepayRestrictions::None => (),
+		};
+
+		Ok(amount.min(current_debt))
+	}
+
+	fn ensure_can_write_off(
+		&self,
+		limit: &WriteOffState<T::Rate>,
+		status: &WriteOffStatus<T::Rate>,
+	) -> DispatchResult {
+		ensure!(
+			T::Time::now().as_secs() > self.info.schedule.maturity.date(),
+			Error::<T>::from(WrittenOffError::MaturityDateNotPassed)
+		);
+
+		ensure!(
+			status.percentage >= limit.percentage && status.penalty >= limit.penalty,
+			Error::<T>::from(WrittenOffError::LessThanPolicy)
+		);
+
+		Ok(())
+	}
+
+	fn ensure_can_close(&self) -> DispatchResult {
+		ensure!(
+			self.normalized_debt.is_zero(),
+			Error::<T>::from(CloseLoanError::NotFullyRepaid)
+		);
+
+		Ok(())
+	}
+
+	pub fn update_time(&mut self, when: Moment) {
+		if when > self.last_updated {
+			self.last_updated = when;
+		}
+	}
+
+	pub fn borrow(&mut self, amount: T::Balance) -> DispatchResult {
+		self.ensure_can_borrow(amount)?;
+
+		self.total_borrowed.ensure_add_assign(amount)?;
+
+		self.normalized_debt = T::InterestAccrual::adjust_normalized_debt(
+			self.info.interest_rate,
+			self.normalized_debt,
+			Adjustment::Increase(amount),
+		)?;
+
+		Ok(())
+	}
+
+	pub fn repay(&mut self, amount: T::Balance) -> Result<T::Balance, DispatchError> {
+		let amount = self.ensure_can_repay(amount)?;
+
+		self.total_repaid.ensure_add_assign(amount)?;
+
+		self.normalized_debt = T::InterestAccrual::adjust_normalized_debt(
+			self.info.interest_rate,
+			self.normalized_debt,
+			Adjustment::Decrease(amount),
+		)?;
+
+		Ok(amount)
+	}
+
+	pub fn write_off(
+		&mut self,
+		limit: &WriteOffState<T::Rate>,
+		new_status: &WriteOffStatus<T::Rate>,
+	) -> DispatchResult {
+		self.ensure_can_write_off(limit, new_status)?;
+
+		let prev_interest_rate = self.info.interest_rate;
+		let next_interest_rate = self.interest_rate_with(new_status.penalty)?;
+
+		T::InterestAccrual::reference_rate(next_interest_rate)?;
+
+		self.normalized_debt = T::InterestAccrual::renormalize_debt(
+			prev_interest_rate,
+			next_interest_rate,
+			self.normalized_debt,
+		)?;
+		self.written_off_status = new_status.clone();
+		self.info.interest_rate = next_interest_rate;
+
+		T::InterestAccrual::unreference_rate(prev_interest_rate)
+	}
```

Comment:
This is the most critical part where the accrual computation is done. More eyes here much better 🙌🏻

Reply:
For the repay method, one question: I think previously, you could submit `repay` with a larger amount than the current debt, and the repay method would implicitly only repay the actual debt. @onnovisser could confirm how exactly the UI integration works. The reason it was used this way is that to close a loan, you need to repay the entire debt. But this changes every block, and it's impossible to predict which block the loan will be repaid in. So either we need to keep the old behaviour or we need to add a `repay_full` extrinsic.

Reply:
It's currently working as before, `ensure_can_repay()` returns the correct amount if the value is higher than current debt

Reply:
Ah I see, perfect! Somehow thought it wasn't the same behaviour. Maybe we should actually add a comment for this still, since it is a bit implicit behaviour.

Reply:
Ok! Probably, `ensure_can_repay` is not the best name if it also returns a new `amount` value. I'll think about it. Thanks!

Reply:
Just realized that Jeroen is off, cc @mustermeiszer @branan

Reply:
That's right, but it's less of an issue: the UI can just only allowing borrowing up to what the debt will be say an hour from now, and this difference is not significant.  While for repaying it actually matters as you have to repay the exact amount in order to close the loan.

cc @onnovisser